corehq/messaging/scheduling/views.py

```diff
@@ -107,6 +117,130 @@ def get(self, request, *args, **kwargs):
+
+    def post(self, request, *args, **kwargs):
+        if self.message_form.is_valid():
+            # TODO editing should not create a new one
+            values = self.message_form.cleaned_data
+            if values['send_frequency'] == 'immediately':
+                if values['translate']:
+                    messages = {}
+                    for lang in self.project_languages:
+                        key = 'message_%s' % lang
+                        if key in values:
+                            messages[lang] = values[key]
+                    content = SMSContent(message=messages)
+                else:
+                    content = SMSContent(message={'*': values['non_translated_message']})
+                schedule = AlertSchedule.create_simple_alert(self.domain, content)
+                broadcast = ImmediateBroadcast(
+                    domain=self.domain, name=values['schedule_name'], schedule=schedule
+                )
+                broadcast.save()
+                recipients = [('CommCareUser', r_id) for r_id in values['recipients']]
+                refresh_alert_schedule_instances.delay(schedule, recipients)
+
+            return HttpResponseRedirect(reverse(BroadcastListView.urlname, args=[self.domain]))
+        return self.get(request, *args, **kwargs)
+
+
+class EditMessageView(CreateMessageView):
+    urlname = 'edit_message'
+    page_title = ugettext_lazy('Edit Message')
+
+    @property
+    def page_url(self):
+        return reverse(self.urlname, args=[self.domain, self.broadcast_id])
+
+    @property
+    def broadcast_id(self):
+        return self.kwargs.get('broadcast_id')
+
+    @cached_property
+    def broadcast(self):
+        try:
+            broadcast = ImmediateBroadcast.objects.prefetch_related('schedule').get(pk=self.broadcast_id)
+        except:
+            raise Http404()
+
+        return broadcast
+
+    @cached_property
+    def message_form(self):
+        if self.request.method == 'POST':
+            return MessageForm(self.request.POST, **self.form_kwargs)
+
+        broadcast = self.broadcast
+        schedule = broadcast.schedule
+        schedule_instances = get_alert_schedule_instances_for_schedule(schedule)
+        recipients = [
+            (instance.recipient_type, instance.recipient_id)
+            for instance in schedule_instances
+        ]
+        initial = {
+            'schedule_name': broadcast.name,
+            'send_frequency': 'immediately',
+            'recipients': recipients,
+            'content': 'sms',
+            'message': schedule.memoized_events[0].content.message,
+        }
+        return MessageForm(initial=initial, **self.form_kwargs)
+
+    def post(self, request, *args, **kwargs):
+        values = self.message_form.cleaned_data
+        # TODO shouldn't have gotten here. add soft assert
+        if values['send_frequency'] == 'immediately':
+            # TODO 404 probably isn't the right thing
+            raise Http404()
```

Comment:
i think we never expect this to happen since we shouldn't be editing the immediate broadcasts, so i'm ok with raising a normal exception and failing hard if this ever happens

Reply:
I changed this to using a soft assert and returning a 400 here: https://github.com/dimagi/commcare-hq/pull/17722/commits/91355a2008e150633c79c9cfeea0ca5fe5f5d8c9

Does that seem like an ok solution or should we 500?

Reply:
i think it'd prefer a 500 since it would really be a bug that we'd need to fix

Reply:
https://github.com/dimagi/commcare-hq/pull/17722/commits/3e0fb84ab5d4333c205561c2c2281fd296c4a655